### PR TITLE
refactor(backup/restore): Add support for CSI based backup/restore

### PR DESCRIPTION
**Pull request by shashank855**
Signed-off-by:  <ranjanshashank855@gmail.com>

**What this PR does / why we need it**:

- Add support for CSI based Backup/Restore.

**Following is the log of ansible test container:**

- https://gist.github.com/shashank855/775a15d645c9e0eb306e306317c15101

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

- https://github.com/openebs/e2e-tests/issues/415

**Checklist**

* [x] Does this PR have a corresponding GitHub issue?
* [ ] Have you included relevant README for the chaoslib/experiment with details?
* [ ] Have you added debug messages where necessary? 
* [ ] Have you added task comments where necessary? 
* [ ] Have you tested the changes for possible failure conditions?
* [ ] Have you provided the positive & negative test logs for the litmusbook execution?
* [ ] Does the litmusbook ensure idempotency of cluster state?, i.e., is cluster restored to original state?
* [ ] Have you used non-shell/command modules for Kubernetes tasks?
* [ ] Have you (jinja) templatized custom scripts that is run by the litmusbook, if any? 
* [ ] Have you (jinja) templatized Kubernetes deployment manifests used by the litmusbook, if any?
* [ ] Have you reused/created util functions instead of repeating tasks in the litmusbook?
* [ ] Do the artifacts follow the appropriate directory structure? 
* [ ] Have you isolated storage (eg: OpenEBS) specific implementations, checks? 
* [ ] Have you isolated platform (eg: baremetal kubeadm/openshift/aws/gcloud) specific implementations, checks?  
* [ ] Are the ansible facts well defined? Is the scope explicitly set for playbook & included utils?
* [ ] Have you ensured minimum/careful usage of shell utilities (awk, grep, sed, cut, xargs etc.,)?
* [ ] Can the limtusbook be executed both from within & outside a container (configurable paths, no hardcode)?
* [ ] Can you suggest the minimal resource requirements for the litmusbook execution?
* [ ] Does the litmusbook job artifact carry comments/default options/range for the ENV tunables?
* [ ] Has the litmusbooks been linted? 

**Special notes for your reviewer**:

### CODE REVIEW

File: experiments/functional/backup_and_restore/cstor_post_restore.yml [hunk 1]
```diff
@@ -6,31 +6,68 @@
...
+        shell: kubectl exec -it {{item.0.stdout}} -n {{ operator }} -- zfs set io.openebs:targetip={{ target_ip.stdout }} {{item.1}}
+        with_together:
+          - "{{ pool_pod.results }}"  
+          - "{{ pool_vol_name.stdout_lines }}"
+
+    when: storage_engine == "cstor"
+  
+  - block: 
+
+      - name: Getting the CVR name from the corresponding PV
```
Comment (nsathyaseelan) [inline experiments/functional/backup_and_restore/cstor_post_restore.yml]: Getting cspi name

File: experiments/functional/backup_and_restore/cstor_post_restore.yml [hunk 2]
```diff
@@ -6,31 +6,68 @@
...
+        shell:  kubectl get cvr -n {{ operator }} -l openebs.io/persistent-volume={{ vol_name }} -o jsonpath='{range.items[*]}{.metadata.labels.cstorpoolinstance\.openebs\.io/name}{"\n"}{end}'
+        register: cvr_name
+   
+      - name: Geting the CSPI name corresponding to CVR
```
Comment (nsathyaseelan) [inline experiments/functional/backup_and_restore/cstor_post_restore.yml]: In this task also we are getting cspi name only

File: experiments/functional/backup_and_restore/cstor_post_restore.yml [hunk 3]
```diff
@@ -6,31 +6,68 @@
...
+      - name: Getting the CVR name from the corresponding PV
+        shell:  kubectl get cvr -n {{ operator }} -l openebs.io/persistent-volume={{ vol_name }} -o jsonpath='{range.items[*]}{.metadata.labels.cstorpool\.openebs\.io/name}{"\n"}{end}'
+        register: cvr_name
    
-  - name: Geting the CSP name corresponding to CVR
-    shell: kubectl get csp {{item}} -n {{ operator }} -o jsonpath='{.metadata.name}'
-    register: csp_pod
-    with_items: "{{ cvr_name.stdout_lines }}"     
+      - name: Geting the CSP name corresponding to CVR
+        shell: kubectl get csp {{item}} -n {{ operator }} -o jsonpath='{.metadata.name}'
```
Comment (gprasath) [inline experiments/functional/backup_and_restore/cstor_post_restore.yml]: namespace not required

**Review by nsathyaseelan — CHANGES_REQUESTED** [review]

File: experiments/functional/backup_and_restore/cstor_post_restore.yml (hunk 2 above)
Comment (shashank855) [reply experiments/functional/backup_and_restore/cstor_post_restore.yml]: here we are getting CSPI name only

File: experiments/functional/backup_and_restore/cstor_post_restore.yml (hunk 1 above)
Comment (shashank855) [reply experiments/functional/backup_and_restore/cstor_post_restore.yml]: Getting CVR name in this task.

Comment (nsathyaseelan) [reply experiments/functional/backup_and_restore/cstor_post_restore.yml]: No @shashank855  here you are getting cspi name as a list can you verify

Comment (shashank855) [reply experiments/functional/backup_and_restore/cstor_post_restore.yml]: Thanks @nsathyaseelan  Got It.. Made the changes

**Review by nsathyaseelan — APPROVED** [review]
lgtm

File: experiments/functional/backup_and_restore/cstor_post_restore.yml [hunk 4]
```diff
@@ -6,31 +6,63 @@
...
-    register: cvr_name
+  - block:
+
+      - name: Getting the CVR name from the corresponding PV
+        shell:  kubectl get cvr -n {{ operator }} -l openebs.io/persistent-volume={{ vol_name }} -o jsonpath='{range.items[*]}{.metadata.labels.cstorpool\.openebs\.io/name}{"\n"}{end}'
+        register: cvr_name
+   
+      - name: Geting the CSP name corresponding to CVR
+        shell: kubectl get csp {{item}}  -o jsonpath='{.metadata.name}'
+        register: csp_pod
```
Comment (gprasath) [inline experiments/functional/backup_and_restore/cstor_post_restore.yml]: where this variable is used. if not necessary, remove this task

File: experiments/functional/backup_and_restore/cstor_post_restore.yml [hunk 5]
```diff
@@ -6,31 +6,63 @@
     delay: 5
     retries: 10
 
-  - name: Getting the CVR name from the corresponding PV
-    shell:  kubectl get cvr -n {{ operator }} -l openebs.io/persistent-volume={{ vol_name }} -o jsonpath='{range.items[*]}{.metadata.labels.cstorpool\.openebs\.io/name}{"\n"}{end}'
-    register: cvr_name
+  - block:
+
+      - name: Getting the CVR name from the corresponding PV
+        shell:  kubectl get cvr -n {{ operator }} -l openebs.io/persistent-volume={{ vol_name }} -o jsonpath='{range.items[*]}{.metadata.labels.cstorpool\.openebs\.io/name}{"\n"}{end}'
```
Comment (gprasath) [inline experiments/functional/backup_and_restore/cstor_post_restore.yml:12]: csp will be obtained here, not cvr

Comment (shashank855) [reply experiments/functional/backup_and_restore/cstor_post_restore.yml:12]: Made the change.

**Review by gprasath — APPROVED** [review]

**Pull request merged**